grpc/execution/server.go

```diff
@@ -123,11 +129,232 @@ func (s *ExecutionServiceServer) FinalizeBlock(ctx context.Context, req *executi
+// CommitmentState.
+func (s *ExecutionServiceServerV1Alpha2) UpdateCommitmentState(ctx context.Context, req *executionv1a2.UpdateCommitmentStateRequest) (*executionv1a2.CommitmentState, error) {
+	softEthHash := common.BytesToHash(req.CommitmentState.Soft.Hash)
+	firmEthHash := common.BytesToHash(req.CommitmentState.Firm.Hash)
+
+	// Validate that the firm and soft blocks exist before going further
+	softBlock := s.bc.GetBlockByHash(softEthHash)
+	if softBlock == nil {
+		return nil, status.Error(codes.InvalidArgument, "Soft block specified does not exist")
+	}
+	firmBlock := s.bc.GetBlockByHash(firmEthHash)
+	if firmBlock == nil {
+		return nil, status.Error(codes.InvalidArgument, "Firm block specified does not exist")
+	}
+
+	currentHead := s.bc.CurrentBlock().Hash()
+
+	// Update the head block to soft commitment This must be done before last
+	// validation step, we can only check if a block belongs to the canonical
+	// chain.
+	if currentHead != softEthHash {
+		if _, err := s.bc.SetCanonical(softBlock); err != nil {
+			return nil, status.Error(codes.Internal, "could not update head to safe hash")
+		}
+	}
+
+	// Once head is updated validate that firm belongs to chain
+	if rawdb.ReadCanonicalHash(s.eth.ChainDb(), firmBlock.NumberU64()) != firmEthHash {
+		// We don't want partial commitments, rolling back.
+		rollbackBlock := s.bc.GetBlockByHash(currentHead)
+		s.bc.SetCanonical(rollbackBlock)
+
+		return nil, status.Error(codes.InvalidArgument, "Firm block specified does not exist on canonical chain")
+	}
+
+	// Updating the safe and final after everything validated
+	currentSafe := s.bc.CurrentSafeBlock().Hash()
+	if currentSafe != softEthHash {
+		s.bc.SetSafe(softBlock.Header())
+	}
+	currentFirm := s.bc.CurrentFinalBlock().Hash()
+	if currentFirm != firmEthHash {
+		s.bc.SetFinalized(firmBlock.Header())
+	}
+
+	return req.CommitmentState, nil
+}
+
+func (s *ExecutionServiceServerV1Alpha2) getBlockFromIdentifier(identifier *executionv1a2.BlockIdentifier) (*executionv1a2.Block, error) {
+	var header *types.Header
+
+	// Grab the header based on the identifier provided
+	switch id_type := identifier.Identifier.(type) {
+	case *executionv1a2.BlockIdentifier_BlockNumber:
+		header = s.bc.GetHeaderByNumber(uint64(identifier.GetBlockNumber()))
+		break
+	case *executionv1a2.BlockIdentifier_BlockHash:
+		header = s.bc.GetHeaderByHash(common.BytesToHash(identifier.GetBlockHash()))
+		break
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "identifier has unexpected type %T", id_type)
+	}
+
+	if header == nil {
+		return nil, status.Errorf(codes.NotFound, "Couldn't locate block with identifier %s", identifier.Identifier)
+	}
+
+	res, err := s.ethHeaderToExecutionBlock(header)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "internal error")
+	}
+
+	return res, nil
+}
+
+func (s *ExecutionServiceServerV1Alpha2) ethHeaderToExecutionBlock(header *types.Header) (*executionv1a2.Block, error) {
```

Comment:
```suggestion
func ethHeaderToExecutionBlock(header *types.Header) (*executionv1a2.Block, error) {
```

make this a standalone function